### PR TITLE
genimage.bbclass: clean ${B} completely before building

### PR DESCRIPTION
**Pull request by Villemoes**
An alternative to fixing the accumulation of old images reported in
PR #78. Tested lightly with changing combinations of .config files, input images and both, and seems to DTRT.

Signed-off-by: Rasmus Villemoes <rasmus.villemoes@prevas.dk>

### CODE REVIEW

File: classes/genimage.bbclass
Comment (ejoerns) [inline classes/genimage.bbclass:103]: This is not required anymore then as `[cleandirs]` already implies `[dirs]`

File: classes/genimage.bbclass
Comment (ejoerns) [inline classes/genimage.bbclass:106]: This looks much more straightforward than what we had before, yes. Having .config copied in do_genimage only will also automatically deprecate the following `.config.tmp` handling. There we then can simply use `.config`.

**Review by ejoerns — CHANGES_REQUESTED** [review]
This fix for the originally filed issue really looks like the right way to do it.

And it also brings in some useful simplifications.

Thus for merging this, we sould get rid of `[dirs]`, remove additional copying to `.config.tmp` and give the commit message a proper description that does not reference the old PR anymore and simply describes what was changed why.

File: classes/genimage.bbclass
Comment (Villemoes) [reply classes/genimage.bbclass:103]: Yes, in terms of ensuring those directories exist before the task is started (with the additional property of ensuring they are empty). But I don't think it also has the side effect of making the working directory for the task be the last element in cleandirs when dirs is absent. At least that's not how I read the documentation or `bitbake/lib/bb/build.py`.

Comment (Villemoes) [reply classes/genimage.bbclass:103]: Hm, maybe meta-ptx master as-is doesn't need the cwd of do_genimage to be set explicitly to ${B}. But I do need that because I currently work around that using `include` in the main config file got broken when `${S}` and `${B}` were separated by copying all my config fragments to ${B} in a do_genimage_prepend. Until I can rely on `--includepath ${S}` being passed and drop that workaround, ${B} must be the cwd of do_genimage.

Comment (ejoerns) [reply classes/genimage.bbclass:103]: Ok, the cwd would be an argument to keep it.

**Review by ejoerns — APPROVED** [review]
Both the content and the commit message now look excellent to me. :+1:

**Comment by Villemoes** [comment]
> Both the content and the commit message now look excellent to me. +1

Thanks, but it's broken, I left out the source in the sed line. So hold on merging a bit longer :)

**Comment by ejoerns** [comment]
No worries, I waited for the test build anyway and did some local tests in parallel. They already pointed out that 'something' was wrong.

**Comment by ejoerns** [comment]
To be honest, I would probably change this part again after having merged it, as there could be more sed'ing than actually is. I would, for example also like to have a sed for @MACHINE@ as it makes it in various file names an then could be used in genimage config files. For this, I would first copy the file (as originally) and then do some inline sed (-i) commands.

**Comment by Villemoes** [comment]
> To be honest, I would probably change this part again after having merged it, as there could be more sed'ing than actually is. I would, for example also like to have a sed for `@MACHINE@` as it makes it in various file names an then could be used in genimage config files. For this, I would first copy the file (as originally) and then do some inline sed (-i) commands.

Yes, if more replacements are to be done it makes sense to do a copy and then a bunch of `sed -i` so that none of the replacements are treated specially. 

We have a class we use for this kind of thing to avoid error-prone sed replacements; it runs a task between do_patch and do_configure on a given set of files in WORKDIR and replaces @@FOO@@ by the value of `${FOO}`. And we do in fact use that on our config fragments to specify things like U-Boot placement and environment size from bitbake variables.

**Comment by ejoerns** [comment]
> > To be honest, I would probably change this part again after having merged it, as there could be more sed'ing than actually is. I would, for example also like to have a sed for `@MACHINE@` as it makes it in various file names an then could be used in genimage config files. For this, I would first copy the file (as originally) and then do some inline sed (-i) commands.
> 
> Yes, if more replacements are to be done it makes sense to do a copy and then a bunch of `sed -i` so that none of the replacements are treated specially.
> 
> We have a class we use for this kind of thing to avoid error-prone sed replacements; it runs a task between do_patch and do_configure on a given set of files in WORKDIR and replaces @@foo@@ by the value of `${FOO}`. And we do in fact use that on our config fragments to specify things like U-Boot placement and environment size from bitbake variables.

How does it handle changing variable values? This was the issue we had to resolve here, too. `do_unpack()` copies the original file to `${WORKDIR}` and then `do_configure()` (or any later task) did the sed on this file by using the `@@` markers. If then the value changes, it invalidates the task using this variable, but the file in workdir would not have the replace markers anymore and thus the change would have no effect. Did you add the variables as dependencies to do_unpack then explicitly?

**Comment by Villemoes** [comment]
> > > To be honest, I would probably change this part again after having merged it, as there could be more sed'ing than actually is. I would, for example also like to have a sed for `@MACHINE@` as it makes it in various file names an then could be used in genimage config files. For this, I would first copy the file (as originally) and then do some inline sed (-i) commands.
> > 
> > 
> > Yes, if more replacements are to be done it makes sense to do a copy and then a bunch of `sed -i` so that none of the replacements are treated specially.
> > We have a class we use for this kind of thing to avoid error-prone sed replacements; it runs a task between do_patch and do_configure on a given set of files in WORKDIR and replaces @@foo@@ by the value of `${FOO}`. And we do in fact use that on our config fragments to specify things like U-Boot placement and environment size from bitbake variables.
> 
> How does it handle changing variable values? This was the issue we had to resolve here, too. `do_unpack()` copies the original file to `${WORKDIR}` and then `do_configure()` (or any later task) did the sed on this file by using the `@@` markers. If then the value changes, it invalidates the task using this variable, but the file in workdir would not have the replace markers anymore and thus the change would have no effect. Did you add the variables as dependencies to do_unpack then explicitly?

Yup, it's not completely trivial to get right, but yes, adding the variables used to do_unpack[vardeps] is exactly what we do - plus marking the source files via bb.parse.mark_dependency() so that a change in one of them causes the whole thing to be reparsed (otherwise some tasks will run with stale metadata). I started writing that class after I had one too many times copy-pasted a `sed s/@FOO@/${FOO}/` to `sed /@BAR@/${FOO}/` - and the #LOC overhead of adding some unpack[postfunc] just to do a single substitution was quite annoying as well.

**Comment by ejoerns** [comment]
> Yup, it's not completely trivial to get right, but yes, adding the variables used to do_unpack[vardeps] is exactly what we do - plus marking the source files via bb.parse.mark_dependency() so that a change in one of them causes the whole thing to be reparsed (otherwise some tasks will run with stale metadata). I started writing that class after I had one too many times copy-pasted a `sed s/@FOO@/${FOO}/` to `sed /@BAR@/${FOO}/` - and the #LOC overhead of adding some unpack[postfunc] just to do a single substitution was quite annoying as well.

Nice! For geniamge.bbclass I hope having 2-3 sed lines will be sufficient first of all ;)

I just fixed the sstate mirror build, but github actions run will not be able to quickly benefit from this..

As the tests do not (yet) really test something about genimage class and my local test looks ok now, I would then force-merge this if nothing speak against this from your side.

**Comment by Villemoes** [comment]
> > Yup, it's not completely trivial to get right, but yes, adding the variables used to do_unpack[vardeps] is exactly what we do - plus marking the source files via bb.parse.mark_dependency() so that a change in one of them causes the whole thing to be reparsed (otherwise some tasks will run with stale metadata). I started writing that class after I had one too many times copy-pasted a `sed s/@FOO@/${FOO}/` to `sed /@BAR@/${FOO}/` - and the #LOC overhead of adding some unpack[postfunc] just to do a single substitution was quite annoying as well.
> 
> Nice! For geniamge.bbclass I hope having 2-3 sed lines will be sufficient first of all ;)
> 
> I just fixed the sstate mirror build, but github actions run will not be able to quickly benefit from this..
> 
> As the tests do not (yet) really test something about genimage class and my local test looks ok now, I would then force-merge this if nothing speak against this from your side.

AFAICT it's OK now, so please merge.

**Comment by ejoerns** [comment]
Cherry-picked to 'dunfell' branch.

**Pull request merged**
